### ISSUE TITLE
Ways to reduce cost of backprop for long simulations.

### ISSUE THREAD

**Issue by jnsbck** [jaxleyverse/jaxley#762]
I have spent some time thinking about ways to reduce the cost of backprop and I came across the following [article](https://www.emergentmind.com/topics/memory-efficient-backpropagation-through-time).
>Truncated BPTT splits the sequence into segments of length KK and only propagates gradients within windows; the selection of KK trades off resource and bias. An adaptive strategy ([Aicher et al., 2019](https://www.emergentmind.com/papers/1905.07473)) estimates the rate of geometric gradient decay and selects KK at runtime to keep the gradient bias below a user-specified threshold, ensuring reliable convergence and optimal resource allocation throughout training.

I think there are some interesting ideas in there. 

One idea I had on how to implement this – it might be too simple, but it could also just work – is to just replace `lax.scan` in `integrate` with a "truncated scan". Something like the code below that I got from Claude when I asked it to implement a strided scan.

```python
def strided_scan(step_fn, init_carry, xs, stride=10):
    """BPTT with strides."""
    
    def conditional_stop_gradient(i, carry):
        # Stop gradient for all steps except those divisible by stride
        return jax.lax.cond(
            i % stride == 0,
            lambda c: c,  # Keep gradient
            lambda c: lax.stop_gradient(c),  # Stop gradient
            carry
        )
    
    def step_with_index(carry_and_idx, x):
        carry, idx = carry_and_idx
        carry = conditional_stop_gradient(idx, carry)
        new_carry, y = step_fn(carry, x)
        return (new_carry, idx + 1), y
    
    (final_carry, _), ys = lax.scan(
        step_with_index, 
        (init_carry, 0), 
        xs
    )
    
    return final_carry, ys
```

This should only take the gradients w.r.t. every nth step.

EDIT:
or with windows (sketch also thanks to Claude)

```python
def truncated_ode_scan(ode_step, init_state, ts, window_size):
    """
    Scan with truncated backprop for ODE integration.
    Gradients only flow within windows, not across them.
    """
    def windowed_step(state, window_data):
        window_ts = window_data
        # Integrate over this window
        final_state, trajectory = lax.scan(ode_step, state, window_ts)
        # Stop gradient on state between windows, but keep trajectory grads
        return lax.stop_gradient(final_state), trajectory
    
    # Reshape timesteps into windows
    num_steps = len(ts)
    num_windows = num_steps // window_size
    ts_windowed = ts[:num_windows * window_size].reshape(num_windows, window_size, *ts.shape[1:])
    
    final_state, trajectories = lax.scan(windowed_step, init_state, ts_windowed)
    
    # Flatten back
    full_trajectory = trajectories.reshape(num_windows * window_size, *trajectories.shape[2:])
    
    return final_state, full_trajectory
```
